### DISCUSSION TITLE
How do you mount/unmount drives with clifm?

### DISCUSSION THREAD

**Discussion by Docbroke** [leo-arch/clifm#66] [General]
`net` works well with ftp/ssh but for physical drives where mount point is not fixed (like removable drives), `net` is difficult to use. 

clifm has `mp` to show  mountpoints, but there is no command to show unmounted drives, which may be useful. vifm has `media` command to show removable drives which can be mounted by vifm (using udevil). Can we have something like that for clifm?

So what I am suggesting is a something like
```
$ media

1 /media/sdb1 (/dev/sdb1)
2 /dev/sdc1

Choose device to mount/unmount ('q' to quit):
```

**Comment by leo-arch** [comment]
I see. The problem is drive names (like `/dev/sdc1`, which could become `/dev/sdb1` on next boot). I guess I can modify the `mp` command to handle non-mounted drives as well.

**Comment by Docbroke** [comment]
here is a script I am using to mount drives, not very refined but works for now, I call it from within clifm (as plugin).
```
#!/bin/sh

lsblk || doas fdisk -l || doas parted -l
echo "mount/unmount what (e.g., sdb1) ?"
read device
[[ ! $device ]] && exit

case "$1" in
	u|-u) 
		doas umount -l /media/$device
	;;
	*) 
		[[ -d /media/$device ]] || doas mkdir /media/$device
		doas mount -o users,rw /dev/$device /media/$device
		printf "%s\n" "/media/$device" > "$CLIFM_BUS"
	;;
esac
exit 0
# doas clifm /media/$device &&\
# doas umount -l /media/$device
# clifm /media/$device

```

For clifm I think we need two commands, one to facilitate movement to mount point ( which mp is doing at present), and another one for mounting and dismounting drives ( ?md ), which toggles drive between mounted and unmounted state and also cd to the mountpoint if drive is mounted. The second command can use either udevil or udisks, and if they  are not available mount/umount using root privileges.

**Reply by leo-arch** [reply]
I'll start working on this as soon as the open-with function reaches a more or less stable state.

**Comment by leo-arch** [comment]
Hi @Docbroke. Have you tried `net` using UUID (instead of drive name) for the mount command (using the `-U` parameter)? That's how I mount my local drives from within CliFM, no matter what name the kernel assigned to the drive: the UUID won't change.

Example:
`MountCmd=sudo mount -U c97d91c4-6881-4b2f-b64d-e0253yyyXXZ %m`

This is not exactly what `vifm` does, listing non-mounted drives, but it works. You only need to set up your drives once in the net config file.

**Reply by Docbroke** [reply]
For known drives, `net` is fine to use, using UUID is good idea, I just tried that with  my usually unmounted drives. Problem is mounting usb drives.

Another thing I noticed is if `%m` is missing in `MountCmd` or `UnmountCmd` clifm segfaults. `clifm` probably should produce an error output before continuing or failing (if it's critical), segfault is not expected.

**Reply by leo-arch** [reply]
Shouldn't be segfaulting anymore. As to USB drives, `net` does work provided the drive is known. This is how I set up two of my USB drives for `net`:

```sh
[Sandisk USB]
Comment=Sandisk USB drive
Mountpoint="/media/usb"
MountCmd=sudo mount -o gid=1000,fmask=113,dmask=002 -U 5847-xxxx %m
UnmountCmd=sudo umount %m
AutoUnmount=false
AutoMount=false

[Kingston USB]
Comment=Kingston USB drive
Mountpoint="/media/usb2"
MountCmd=sudo mount -t iso9660 -U 2020-10-01-15-xx-yy-zz %m
UnmountCmd=sudo umount %m
AutoUnmount=false
AutoMount=false
```
Both are working fine.

**Reply by Docbroke** [reply]
I actually meant unknown drives. I frequently need to use usb drives of colleagues.

EDIT: setting gid/fmask/dmask is cool idea, I just updated my plugin script for usb drives, now I don't need to use clifm with doas. mount/umount are already set to work without password.

```
#!/bin/sh

case "$1" in
	# u|-u) doas umount -l /media/$device ;;
	u|-u) 
		doas umount -l /media/*
		printf "%s\n" "$HOME" > "$CLIFM_BUS"
	;;
	*) 
		/bin/lsblk || sudo fdisk -l || sudo parted -l
		echo "mount what (e.g., sdb1) ?"
		read device
		[[ ! $device ]] && exit
		[[ -d /media/$device ]] || doas mkdir /media/$device
		# doas mount -o users,rw /dev/$device /media/$device
		doas mount -o gid=1000,fmask=113,dmask=002 /dev/$device /media/$device
		printf "%s\n" "/media/$device" > "$CLIFM_BUS"
	;;
esac
exit 0

```

**Reply by leo-arch** [reply]
Great. Can you confirm it isn't segfaulting anymore?

The thing with unknown drives is this: I either link CliFM to a new library (`libudev`) only for this, which is not optimal (`udev` is Linux-specific), or rather scan `/dev` searching for unmounted drives (even less optimal).

I'll keep working on this. Just need to find out the best approach.

Btw, cool plugin. I'll give it a try. If everything goes fine, you can send a PR to add your plugin.

**Reply by Docbroke** [reply]
No it doesn't segfault anymore.
I think you can use udevil or udisks, like vifm does. I will add one optional dependency.

**Reply by leo-arch** [reply]
I'll take a look at `udevil` and `udisks`. Thanks for the suggestion.

**Reply by Docbroke** [reply]
I have updated plugin script, to make it easier for other users

```
#!/bin/sh

type doas &> /dev/null && super=doas || super=sudo

[[ -z "$1" ]] && echo " USAGE: usb u|-u|m|-m [device|all]"

case "$1" in
	u|-u) 
		if [[ -n "$2"  ]] && [[ -d /media/"$2" ]]; then
			$super umount -l /media/"$2"
			# printf "%s\n" "$HOME" > "$CLIFM_BUS"
		elif [[ $2 = all ]]; then 
			echo "this will unmount all devices mounted in /media (y/n)"
			read reply
			[[ $reply = y ]] &&\
			$super umount -l /media/*
			# printf "%s\n" "$HOME" > "$CLIFM_BUS"
		else
			lsblk || $super fdisk -l || $super parted -l
			echo "unmount what (e.g., sdb1) ?"
			read device
			[[ ! $device ]] && exit
			doas umount -l /media/"$device"
			# printf "%s\n" "$HOME" > "$CLIFM_BUS"
		fi
	;;
	m|-m) 
		if [[ -n "$2" ]] && [[ -e /dev/"$2" ]]; then
			export device="$2"
		else  
			lsblk || $super fdisk -l || $super parted -l
			echo "mount what (e.g., sdb1) ?"
			read device
		fi

		[[ ! $device ]] && exit
		[[ -d /media/$device ]] || doas mkdir /media/$device
		doas mount -o gid=1000,fmask=113,dmask=002 /dev/$device /media/$device
		printf "%s\n" "/media/$device" > "$CLIFM_BUS"
	;;
esac
exit 0


```

**Reply by leo-arch** [reply]
Thanks! Still couldn't try it. Working on a first version of an improved `mp` command, able to list and mount unknown devices using `udisks2`.

**Reply by leo-arch** [reply]
Done. Give the new `mp` command a try and let me know what you think. The implementation is absolute crap, but works (hope so).

EDIT: It only supports `udisks2` right now, but if everything goes fine it should be a breeze to add support for `udevil` as well. In that case, a new option in the config file should allow the user to choose between these two.

**Comment by Docbroke** [comment]
Works, but list of unmounted drives is too long, it includes /dev/ram /dev/loop etc.

**Reply by leo-arch** [reply]
Great! Let me see how to handle those extra "drives".

I do not have any of those files. `mp` should only list block devices in `/dev` ending with a number (that is, partitions). Are `/dev/ram` and `/dev/loop` block devices in your system?

EDIT: Yeap. Just check in Alpine: all those files are block devices. I'll try filtering whatever doesn't seem to be a storage device. This is why I told you the implementation was crap. I need to find out a better way of getting storage device names.

**Reply by leo-arch** [reply]
Loop and ram block devices should be skipped now.

**Reply by Docbroke** [reply]
list of devices are now as expected.
I think, workflow can be improved. There is no need for `u` flag. Instead if selected drive is mounted it is unmounted, and if it unmounted it gets mounted. simple toggling of mount status shall suffice.

EDIT: and may be on unmounting CD back to previous directory or home directory

**Reply by leo-arch** [reply]
I'll see how to improve the workflow.

Thanks for your suggestions: they feel just natural and consistent. However, doing that implies redefining the whole purpose of the `mp` command: it was originally designed to easily cd into a mountpoint (though I'm not sure how much useful is this; bookmarks are there for a reason). I just do not feel comfortable breaking consistency with previous versions.

Maybe I should rename this new `mp` command to something else (perhaps `media`, just like in `Vifm`), and keep the old `mp` as it was.

**Reply by leo-arch** [reply]
When listing unmounted drives, the new `mp` command is supposed to print, next to drives names, their label, if any. It does this by scanning `/dev/disks/by-label`, but this directory doesn't seem to be available on Alpine. Once more, the current implementation is crap because it is not portable.

**Reply by leo-arch** [reply]
Done. `mp` now behaves as before, while the new feature was moved to the `media` command, doing exactly what you suggested: the `u` flag was removed, meaning that if a device is mounted it will be unmounted, and the other way round. Also, if the current directory is inside the mountpoint, CliFM will cd into the previously visited directory (or $HOME is no previous directory) before unmounting the corresponding device.

**Reply by Docbroke** [reply]
This looks good, `mp` to move and `media` to mount/unmount.

**Reply by leo-arch** [reply]
Added support for `udevil`. By default CliFM uses `udisks2`. To switch to `udevil` use the `--mnt-udevil` command line option.

**Reply by Docbroke** [reply]
If only `udevil` is available in the system and `udisks2` is not available, shouln't `clifm` fallback to `udevil` for media command.

**Reply by leo-arch** [reply]
Absolutely. A simple check should make it.

**Reply by Docbroke** [reply]
thanks for adding `udevil` support, it needs less resources compared to `udisks2`
only problem with `udevil` is it may fail to mount exfat drives, due to problem with exfat-utils, but solution is simple, one has to remove `nonempty` option from `default-options-exfat` in `udevil.conf` https://github.com/IgnorantGuru/udevil/issues/89

**Reply by leo-arch** [reply]
And the difference is huge: 300Kb of `udevil` against 13Mb(!) of `udisks2`. From the Gentoo wiki:
> It can be used with or without systemd, ConsoleKit, policykit, D-Bus, udisks, gvfs and FUSE

 In this case `udevil` should be made the default value. I'll add a note about the exfat issue to the wiki.

**Reply by leo-arch** [reply]
Done. `udevil` is the new default.
